library/x509_crt.c

```diff
@@ -957,8 +957,13 @@ static int x509_get_crt_ext( unsigned char **p,
         if( ret != 0 )
         {
             /* Give the callback (if any) a chance to handle the extension */
-            if( cb != NULL && cb( crt, &extn_oid, is_critical, p, end_ext_octet ) == 0 )
+            if( cb != NULL ) {
+                ret = cb( crt, &extn_oid, is_critical, *p, end_ext_octet );
+                if ( ret != 0 )
```

Comment:
what about returning only if the extension is critical? I mean, changing the if condition to

`if( ret != 0 && is_critical != 0 )`

then there should be no problem because

- if the callback fails on a critical extension, the parsing fails altogher as it must do anyway (the whole point of critical extensions)
- if the callback fails on a non-critical extension, it just gets skipped like before

Reply:
Yes, that was the original behavior and I think it makes more sense.

Reply:
Implemented in 565b52bb727a81b82ad07a9bcca5ca033b554a24